src/htm/algorithms/TemporalMemory.cpp

```diff
@@ -185,45 +185,26 @@ void TemporalMemory::growSynapses_(
-  // It's possible to optimize this, swapping candidates to the end as
-  // they're used. But this is awkward to mimic in other
-  // implementations, especially because it requires iterating over
-  // the existing synapses in a particular order.
-
+  
   vector<CellIdx> candidates(prevWinnerCells.begin(), prevWinnerCells.end());
   NTA_ASSERT(std::is_sorted(candidates.begin(), candidates.end()));
 
-  // Skip cells that are already synapsed on by this segment
-  // Biological motivation (?):
-  // There are structural constraints on the shapes of axons & synapses 
-  // which prevent a large number duplicate of connections.
-  //
-  // It's important to prevent cells from growing duplicate synapses onto a segment, 
-  // because otherwise a strong input would be sampled many times and grow many synapses.
-  // That would give such input a stronger connection. 
-  // Synapses are supposed to have binary effects (0 or 1) but duplicate synapses give 
-  // them (synapses 0/1) varying levels of strength.
-  for (const Synapse& synapse : connections.synapsesForSegment(segment)) {
-    const CellIdx presynapticCell = connections.dataForSynapse(synapse).presynapticCell;
-    const auto already = std::lower_bound(candidates.cbegin(), candidates.cend(), presynapticCell);
-    if (already != candidates.cend() && *already == presynapticCell) {
-      candidates.erase(already);
-    }
-  }
-
+  //figure the number of new synapses to grow
   const size_t nActual = std::min(static_cast<size_t>(nDesiredNewSynapses), candidates.size());
-
-  // Check if we're going to surpass the maximum number of synapses.
+  // ..Check if we're going to surpass the maximum number of synapses.
   Int overrun = static_cast<Int>(connections.numSynapses(segment) + nActual - maxSynapsesPerSegment_);
   if (overrun > 0) {
-    connections.destroyMinPermanenceSynapses(segment, static_cast<Int>(overrun), prevWinnerCells);
+    connections.destroyMinPermanenceSynapses(segment, static_cast<Int>(overrun), prevWinnerCells); //TODO move this functionality to Connections.createSynapse() 
   }
-
-  // Recalculate in case we weren't able to destroy as many synapses as needed.
+  // ..Recalculate in case we weren't able to destroy as many synapses as needed.
   const size_t nActualWithMax = std::min(nActual, static_cast<size_t>(maxSynapsesPerSegment_) - connections.numSynapses(segment));
 
   // Pick nActual cells randomly.
-  for (const auto syn : rng_.sample(candidates, static_cast<UInt>(nActualWithMax))) {
+  rng_.shuffle(candidates.begin(), candidates.end());
+  const size_t nDesired = connections.numSynapses(segment) + nActualWithMax; //num synapses on seg after this function (+-), see #COND
+  for (const auto syn : candidates) {
+    // #COND: this loop finishes two folds: a) we ran out of candidates (above), b) we grew the desired number of new synapses (below)
+    if(connections.numSynapses(segment) == nDesired) break;
```

Comment:
please review this logic. I think it's the same as former: have `candidates`, prune candidates that have duplacation on presyn. See comment above, this fixes a bug in the former implementation.